single_kernel_mongo/managers/ldap.py

```diff
@@ -0,0 +1,162 @@
+        self.certificate_transfer = CertificateTransferRequires(self.charm, self.cert_relation_name)
+
+    def assert_pass_hook_checks(self) -> None:
+        """Runs some hook checks before allowing the hook to run."""
+        if not self.state.db_initialised:
+            raise DeferrableFailedHookChecksError("DB is not initialised")
+        if self.state.is_role(MongoDBRoles.SHARD):
+            self.charm.status_manager.to_blocked("Cannot integrate LDAP with shard.")
+            raise NonDeferrableFailedHookChecksError("Cannot integrate LDAP with shard.")
+        if self.state.upgrade_in_progress:
+            raise DeferrableFailedHookChecksError(
+                "Adding LDAP is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+        return
+
+    def is_valid_ldap_integration(self) -> bool:
+        """Returns True if the integration to ldap is valid."""
+        return (self.state.ldap_relation is not None) and not self.state.is_role(MongoDBRoles.SHARD)
+
+    def on_ldap_ready(self, relation: Relation) -> None:
+        """Runs when LDAP is ready."""
+        self.assert_pass_hook_checks()
+
+        self.charm.status_manager.to_maintenance("Configuring LDAP")
+
+        ldap_data = self.ldap_requirer.consume_ldap_relation_data(relation=relation)
+        if not ldap_data:
+            logger.info("Waiting for LDAP data.")
+            raise WaitingForLdapDataError("Waiting for LDAP data.")
+
+        if not ldap_data.ldaps_urls:
+            logger.info("You must configure Glauth with LDAPS support")
+            raise LDAPSNotEnabledError("LDAPS not enabled.")
+
+        if self.charm.unit.is_leader():
+            self.state.ldap.set_from(ldap_data)
+
+        if self.state.ldap.is_ready():
+            logger.info("Restarting mongodb server for LDAP integration")
+            self.dependent.restart_charm_services()
+
+            self.charm.status_manager.to_active()
+
+    def on_ldap_unavailable(self) -> None:
+        """Runs when the LDAP integration is broken."""
+        self.state.ldap.clean_databag()
+
+        self.dependent.restart_charm_services()
+
+    def on_certificate_available(self, certificate: str, ca: str, chain: list[str]):
+        """Runs when we receive the LDAP certificates."""
+        self.assert_pass_hook_checks()
+        self.state.ldap.set_certificates(certificate, ca, chain)
+
+        full_chain = "\n".join(chain)
+        self.dependent.save_ca_cert_to_trust_store(TrustStoreFiles.LDAP, full_chain)
+
+        if self.state.ldap.is_ready():
+            logger.info("Restarting mongodb server for LDAP integration")
+            self.dependent.restart_charm_services()
+
+    def on_certificate_removed(self):
+        """Runs when the certificate is removed."""
+        self.state.ldap.set_certificates(None, None, None)
+        self.dependent.remove_ca_cert_from_trust_store(TrustStoreFiles.LDAP)
+
+        self.dependent.restart_charm_services()
```

Comment:
If we are in one of the cases of not having the cert and having the ldap interface (and vice versa) is the charm in the broken state? i.e. can't proceed with other events? Or is it only blocking clients?

Reply:
It will not be in a broken state, it can process other events.
However identity users will not be able to log in as ldap won't be configured on the services.